File: app/src/main/java/com/dumuzeyn/mp3player/PlayerWidgetProvider.java
package com.dumuzeyn.mp3player;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.widget.RemoteViews;
import androidx.annotation.Nullable;
import androidx.media3.common.MediaItem;
import androidx.media3.common.Player;
import androidx.media3.session.MediaController;
import androidx.media3.session.SessionToken;
import com.google.common.util.concurrent.ListenableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

public final class PlayerWidgetProvider extends AppWidgetProvider {
    private static final String ACTION_PREFIX = "com.dumuzeyn.mp3player.widget.";
    private static final String ACTION_PREVIOUS = ACTION_PREFIX + "PREVIOUS";
    private static final String ACTION_TOGGLE = ACTION_PREFIX + "TOGGLE";
    private static final String ACTION_NEXT = ACTION_PREFIX + "NEXT";
    private static final String ACTION_REFRESH = ACTION_PREFIX + "REFRESH";
    private static final ExecutorService ARTWORK_EXECUTOR =
            Executors.newSingleThreadExecutor(runnable -> {
                Thread thread = new Thread(runnable, "widget-artwork");
                thread.setDaemon(true);
                return thread;
            });
    private static final AtomicInteger ARTWORK_GENERATION = new AtomicInteger();
    private static final Executor MAIN_EXECUTOR = command ->
            new Handler(Looper.getMainLooper()).post(command);

    @Override
    public void onUpdate(Context context, AppWidgetManager manager, int[] appWidgetIds) {
        connect(context, ACTION_REFRESH);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
        String action = intent.getAction();
        if (action != null && action.startsWith(ACTION_PREFIX)) {
            connect(context, action);
        }
    }

    static void updateFromPlayer(Context context, Player player) {
        MediaItem item = player.getCurrentMediaItem();
        CharSequence title = item == null || item.mediaMetadata.title == null
                ? "Voltune" : item.mediaMetadata.title;
        CharSequence artist = item == null || item.mediaMetadata.artist == null
                ? "" : item.mediaMetadata.artist;
        Uri source = item == null || item.localConfiguration == null
                ? null : item.localConfiguration.uri;
        boolean playing = player.isPlaying();
        int generation = ARTWORK_GENERATION.incrementAndGet();
        render(context.getApplicationContext(), title, artist, playing, null);
        if (source != null) {
            ARTWORK_EXECUTOR.execute(() -> {
                Bitmap artwork = readArtwork(context, source);
                if (generation == ARTWORK_GENERATION.get()) {
                    render(context.getApplicationContext(), title, artist,
                            playing, artwork);
                }
            });
        }
    }

    private static void connect(Context context, String action) {
        Context app = context.getApplicationContext();
        SessionToken token = new SessionToken(app,
                new ComponentName(app, Media3PlayerService.class));
        ListenableFuture<MediaController> future =
                new MediaController.Builder(app, token).buildAsync();
        future.addListener(() -> {
            MediaController controller = null;
            try {
                controller = future.get();
                if (ACTION_PREVIOUS.equals(action)) {
                    controller.seekToPreviousMediaItem();
                } else if (ACTION_TOGGLE.equals(action)) {
                    if (controller.isPlaying()) {
                        controller.pause();
                    } else {
                        controller.play();
                    }
                } else if (ACTION_NEXT.equals(action)) {
                    controller.seekToNextMediaItem();
                }
                updateFromPlayer(app, controller);
            } catch (Exception ignored) {
                render(app, "Voltune", "", false, null);
            } finally {
                if (controller != null) {
                    controller.release();
                }
            }
        }, MAIN_EXECUTOR);
    }

    private static void render(Context context, CharSequence title, CharSequence artist,
            boolean playing, @Nullable Bitmap artwork) {
        AppWidgetManager manager = AppWidgetManager.getInstance(context);
        int[] ids = manager.getAppWidgetIds(new ComponentName(context,
                PlayerWidgetProvider.class));
        if (ids.length == 0) {
            return;
        }
        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.player_widget);
        views.setTextViewText(R.id.widget_title, title);
        views.setTextViewText(R.id.widget_artist, artist);
        views.setImageViewResource(R.id.widget_toggle,
                playing ? R.drawable.ic_widget_pause : R.drawable.ic_widget_play);
        if (artwork == null) {
            views.setImageViewResource(R.id.widget_artwork,
                    R.drawable.voltune_icon_master);
        } else {
            views.setImageViewBitmap(R.id.widget_artwork, artwork);
        }
        views.setOnClickPendingIntent(R.id.widget_root, activityIntent(context));
        views.setOnClickPendingIntent(R.id.widget_previous,
                actionIntent(context, ACTION_PREVIOUS, 1));
        views.setOnClickPendingIntent(R.id.widget_toggle,
                actionIntent(context, ACTION_TOGGLE, 2));
        views.setOnClickPendingIntent(R.id.widget_next,
                actionIntent(context, ACTION_NEXT, 3));
        manager.updateAppWidget(ids, views);
    }

    private static PendingIntent activityIntent(Context context) {
        return PendingIntent.getActivity(context, 0, new Intent(context, MainActivity.class),
                PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT);
    }

    private static PendingIntent actionIntent(Context context, String action, int requestCode) {
        Intent intent = new Intent(context, PlayerWidgetProvider.class).setAction(action);
        return PendingIntent.getBroadcast(context, requestCode, intent,
                PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT);
    }

    @Nullable
    private static Bitmap readArtwork(Context context, Uri source) {
        MediaMetadataRetriever retriever = new MediaMetadataRetriever();
        try {
            retriever.setDataSource(context, source);
            byte[] bytes = retriever.getEmbeddedPicture();
            if (bytes == null || bytes.length > 8 * 1024 * 1024) {
                return null;
            }
            Bitmap bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
            if (bitmap == null) {
                return null;
            }
            int max = Math.max(bitmap.getWidth(), bitmap.getHeight());
            if (max <= 320) {
                return bitmap;
            }
            float scale = 320f / max;
            return Bitmap.createScaledBitmap(bitmap, Math.round(bitmap.getWidth() * scale),
                    Math.round(bitmap.getHeight() * scale), true);
        } catch (Exception ignored) {
            return null;
        } finally {
            try {
                retriever.release();
            } catch (Exception ignored) {
                // Nothing else owns the retriever.
            }
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/PlayerWidgetProvider.kt
package com.dumuzeyn.mp3player

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.widget.RemoteViews
import androidx.media3.common.Player
import androidx.media3.session.MediaController
import androidx.media3.session.SessionToken
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicInteger
import kotlin.math.max
import kotlin.math.roundToInt

class PlayerWidgetProvider : AppWidgetProvider() {
    override fun onUpdate(
        context: Context,
        manager: AppWidgetManager,
        appWidgetIds: IntArray,
    ) {
        connect(context, ACTION_REFRESH)
    }

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        val action = intent.action
        if (action != null && action.startsWith(ACTION_PREFIX)) connect(context, action)
    }

    companion object {
        private const val ACTION_PREFIX = "com.dumuzeyn.mp3player.widget."
        private const val ACTION_PREVIOUS = ACTION_PREFIX + "PREVIOUS"
        private const val ACTION_TOGGLE = ACTION_PREFIX + "TOGGLE"
        private const val ACTION_NEXT = ACTION_PREFIX + "NEXT"
        private const val ACTION_REFRESH = ACTION_PREFIX + "REFRESH"
        private val artworkExecutor = Executors.newSingleThreadExecutor { runnable ->
            Thread(runnable, "widget-artwork").apply { isDaemon = true }
        }
        private val artworkGeneration = AtomicInteger()
        private val mainExecutor = Executor { command ->
            Handler(Looper.getMainLooper()).post(command)
        }

        @JvmStatic
        fun updateFromPlayer(context: Context, player: Player) {
            val item = player.currentMediaItem
            val title: CharSequence = item?.mediaMetadata?.title ?: "Voltune"
            val artist: CharSequence = item?.mediaMetadata?.artist ?: ""
            val source = item?.localConfiguration?.uri
            val playing = player.isPlaying
            val generation = artworkGeneration.incrementAndGet()
            render(context.applicationContext, title, artist, playing, null)
            if (source != null) {
                artworkExecutor.execute {
                    val artwork = readArtwork(context, source)
                    if (generation == artworkGeneration.get()) {
                        render(context.applicationContext, title, artist, playing, artwork)
                    }
                }
            }
        }

        private fun connect(context: Context, action: String) {
            val app = context.applicationContext
            val token = SessionToken(app, ComponentName(app, Media3PlayerService::class.java))
            val future = MediaController.Builder(app, token).buildAsync()
            future.addListener(
                {
                    var controller: MediaController? = null
                    try {
                        controller = future.get()
                        when (action) {
                            ACTION_PREVIOUS -> controller.seekToPreviousMediaItem()
                            ACTION_TOGGLE -> if (controller.isPlaying) {
                                controller.pause()
                            } else {
                                controller.play()
                            }
                            ACTION_NEXT -> controller.seekToNextMediaItem()
                        }
                        updateFromPlayer(app, controller)
                    } catch (_: Exception) {
                        render(app, "Voltune", "", false, null)
                    } finally {
                        controller?.release()
                    }
                },
                mainExecutor,
            )
        }

        private fun render(
            context: Context,
            title: CharSequence,
            artist: CharSequence,
            playing: Boolean,
            artwork: Bitmap?,
        ) {
            val manager = AppWidgetManager.getInstance(context)
            val ids = manager.getAppWidgetIds(
                ComponentName(context, PlayerWidgetProvider::class.java),
            )
            if (ids.isEmpty()) return
            val views = RemoteViews(context.packageName, R.layout.player_widget)
            views.setTextViewText(R.id.widget_title, title)
            views.setTextViewText(R.id.widget_artist, artist)
            views.setImageViewResource(
                R.id.widget_toggle,
                if (playing) R.drawable.ic_widget_pause else R.drawable.ic_widget_play,
            )
            if (artwork == null) {
                views.setImageViewResource(R.id.widget_artwork, R.drawable.voltune_icon_master)
            } else {
                views.setImageViewBitmap(R.id.widget_artwork, artwork)
            }
            views.setOnClickPendingIntent(R.id.widget_root, activityIntent(context))
            views.setOnClickPendingIntent(
                R.id.widget_previous,
                actionIntent(context, ACTION_PREVIOUS, 1),
            )
            views.setOnClickPendingIntent(
                R.id.widget_toggle,
                actionIntent(context, ACTION_TOGGLE, 2),
            )
            views.setOnClickPendingIntent(
                R.id.widget_next,
                actionIntent(context, ACTION_NEXT, 3),
            )
            manager.updateAppWidget(ids, views)
        }

        private fun activityIntent(context: Context): PendingIntent = PendingIntent.getActivity(
            context,
            0,
            Intent(context, MainActivity::class.java),
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT,
        )

        private fun actionIntent(
            context: Context,
            action: String,
            requestCode: Int,
        ): PendingIntent = PendingIntent.getBroadcast(
            context,
            requestCode,
            Intent(context, PlayerWidgetProvider::class.java).setAction(action),
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT,
        )

        private fun readArtwork(context: Context, source: Uri): Bitmap? {
            val retriever = MediaMetadataRetriever()
            try {
                retriever.setDataSource(context, source)
                val bytes = retriever.embeddedPicture
                if (bytes == null || bytes.size > 8 * 1024 * 1024) return null
                val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size) ?: return null
                val largestSide = max(bitmap.width, bitmap.height)
                if (largestSide <= 320) return bitmap
                val scale = 320f / largestSide
                return Bitmap.createScaledBitmap(
                    bitmap,
                    (bitmap.width * scale).roundToInt(),
                    (bitmap.height * scale).roundToInt(),
                    true,
                )
            } catch (_: Exception) {
                return null
            } finally {
                try {
                    retriever.release()
                } catch (_: Exception) {
                }
            }
        }
    }
}
